indico/modules/events/registration/fields/accompanying_test.py

```diff
@@ -0,0 +1,243 @@
+# This file is part of Indico.
+# Copyright (C) 2002 - 2022 CERN
+#
+# Indico is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see the
+# LICENSE file for more details.
+
+import pytest
+from marshmallow import ValidationError
+
+from indico.modules.events.features.util import set_feature_enabled
+from indico.modules.events.registration.models.form_fields import RegistrationFormField
+from indico.modules.events.registration.models.forms import RegistrationForm
+from indico.modules.events.registration.models.items import RegistrationFormSection
+from indico.modules.events.registration.models.registrations import Registration, RegistrationData, RegistrationState
+
+
+pytest_plugins = 'indico.modules.events.registration.testing.fixtures'
+
+
+@pytest.fixture
+def create_accompanying_persons_field(db, dummy_regform):
+    def _create_accompanying_persons_field(max_persons, persons_count_against_limit,
+                                           registration=None, data=None, no_of_persons=0):
+        section = RegistrationFormSection(
+            registration_form=dummy_regform,
+            title='dummy_section',
+            is_manager_only=False
+        )
+        db.session.add(section)
+        db.session.flush()
+        field = RegistrationFormField(
+            input_type='accompanying_persons',
+            title='Field',
+            parent_id=section.id,
+            registration_form=dummy_regform
+        )
+        field.field_impl.form_item.data = {
+            'max_persons': max_persons,
+            'persons_count_against_limit': persons_count_against_limit,
+        }
+        field.versioned_data = field.field_impl.form_item.data
+        if registration:
+            RegistrationData(
+                registration_id=registration.id,
+                field_data=field.current_data,
+                data=data if data is not None else _create_accompanying_persons(no_of_persons)
+            )
+            db.session.expire(registration)
```

Comment:
This was a workaround to force the `registration.data` cache to be reset. Not sure if this is the best solution...

Reply:
The reason why you need this is a very common mistake with sqlalchemy relationships: Instead of using the relationships, you use the FKs, but FK columns almost never need to be assigned a value directly. If you used `registration=registration` instead of `registration_id=registration.id` a few lines above, it would most likely work as expected.

Or, a way that's even cleaner IMHO:

```py
registration.data = RegistrationData(...)
```

In that case you don't have to pass the registration as a kwarg at all, because assigning it to the other side of the relationship will do everything that's needed. And it's also very clear that you are setting the registration's data here, while with the `RegistrationData()` that's not assigned to anything it's much less explicit.

Reply:
```py
        if registration:
            registration.data.append(RegistrationData(
                field_data=field.current_data,
                data=(data if data is not None else _create_accompanying_persons(num_persons))
            ))
            db.session.flush()
```